keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+- A dual-stack capable CNI plugin should return pod IP assignments to kubelet via the "ips" array defined in the [CNI plugin result specification](https://github.com/containernetworking/cni/blob/master/SPEC.md#result).
+- It's expected that plugins will typically not assign more than one IPv4 address per pod.
+- The order of assigned IP addresses that are returned in the CNI plugin result for pod creation will be used to determine the "Default Pod IP" as described in the previous section.
+
+### Support for Multiple Service IPs per Service
+
+This proposal adds the capability to define multiple service CIDRs for a cluster.
+
+#### First Release of Dual Stack: No Service Address Configuration "Knobs"
+For the first release of Kubernetes dual-stack support, no new configuration "knobs" will be added for service definitions. This greatly simplifies the design and implementation, but requires imposing the following behavior:
+- Service IP allocation: Kubernetes will always allocate a service IP from each service CIDR for each service that is created. (In the future, we might want to consider adding configuration options to allow a user to select e.g. whether a given service should be assigned only IPv4, only IPv6, or both IPv4 and IPv6 service IPs.)
+- Service mappings: Kubernetes will make the following service mappings (e.g. via Kube-Proxy) between service IPs and backend pod IPs:
+  - Each IPv4 service IP will be mapped to all IPv4 addresses on backend pods
+  - Each IPv6 service IP will be mapped to all IPv6 addresses on backend pods
+
+(Note that this assumed behavior works very well for the typical dual-stack case where there are one IPv4 and one IPv6 service CIDR, and one IPv4 and one IPv6 address per pod.)
+
+#### Versioned API Change: Service v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal adds a new "ExtraClusterIPs" field to the core V1 version of the [Service V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#service-v1-core), while maintaining the existing "ClusterIP" field. This change is in accordance with the [Kubernetes API change
+quidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md):
+
+```
+    // clusterIP is the IP address of the service and is usually assigned
+    // randomly by the master.
+    ClusterIP string `json:"clusterIP,omitempty" protobuf:"bytes,3,opt,name=clusterIP"`
+
+    // Additional cluster IPs
+    ExtraClusterIPs []string `json:"extraclusterIP,omitempty" protobuf:"bytes,3,opt,name=ExtraclusterIPs"`
+```
+
+#### Internal Representation:
+The internal representation for the service IP (a.k.a. cluster IP) will be changed from a singular string to a slice of strings, and renamed "ClusterIPs":
+```
+    ClusterIPs []string `json:"clusterIPs,omitempty" protobuf:"bytes,3,opt,name=clusterIPs"`
```

Comment:
I think you can use this construct in the external struct, too. Just require that ClusterIP == ClusterIPs[0]. Convert and default in a few places, and this is not a terrible user experience.

Reply:
Ah, I see there's already discussion about that for pod IPs--if cluster IP is already mutable then what I said probably won't work.